### ISSUE TITLE
Bug: New lua scripts fail to get values in self.Properties 

### ISSUE THREAD

**Issue by galibzon** [o3de/o3de#10076]
**Describe the bug**
In game mode, new lua scripts get nil values within the `Properties`  table.

**Assets required**
Here is a test level that contains all the pieces: [spawnable_with_lua_property_test.zip](https://github.com/o3de/o3de/files/8886002/spawnable_with_lua_property_test.zip)

**Steps to reproduce**
Use `AutomatedTesting` project, or any other game project.
Unzip the attached level under the `Levels` folder.
In the Editor Load the attached level called `spawnable_with_lua_property_test`.
Enter game mode.

**Expected behavior**
In game mode, there are no lua errors and the Shader Ball is relocated according to the value in `self.Properties.destination`.

**Actual behavior**
In game mode, I get the following errors:
```

<09:03:34> Disable Accelerators
<09:03:34> Entered game mode
<09:03:34> [Warning] (Script) - Property destination not found in entity table. Please push this property to your slice to avoid decrease in performance.
 ---------- START STACK TRACE ---------- 
=[C](-1): index
@levels/spawnable_with_lua_property_test/object_relocator.lua(11): ?

 ---------- END STACK TRACE ----------
<09:03:34> (Script) - object_relocator:OnActivate destinationnil
<09:03:34> [Warning] (Script) - Property destination not found in entity table. Please push this property to your slice to avoid decrease in performance.
 ---------- START STACK TRACE ---------- 
=[C](-1): index
@levels/spawnable_with_lua_property_test/object_relocator.lua(12): ?

 ---------- END STACK TRACE ----------
<09:03:34> (Script) - object_relocator:OnTick will reposition to nil
<09:03:34> [Error] (Script) - Lua failed to call method: cannot convert parameter 2 from nil to const Vector3&
 ---------- START STACK TRACE ---------- 
=[C](-1): SetWorldTranslation
@levels/spawnable_with_lua_property_test/object_relocator.lua(27): ?

 ---------- END STACK TRACE ----------
<09:04:01> Enable Accelerators
<09:04:01> (System) - > disconnect : <empty>
<09:04:01> Exited game mode
```

**Screenshots/Video**
N/A

**Found in Branch**
`development` branch at:
commit 47a2dc800d42315a43645d4511493d1f7e43fb5c (HEAD -> development, upstream/development)
Date:   Fri Jun 10 19:54:05 2022 -0500

**Desktop/Device (please complete the following information):**
 - Device: [PC] 
 - OS: [Windows]
 - Version [10]
 - CPU [AMD Ryzen Threadripper 3970X 32-Core Processor 3.70 GHz ]
 - GPU [NVidia RTX 3070]
 - Memory [e.g. 64GB]

**Additional context**
After changing the scrip `Properties` several times, and resaving the levels, the bug doesn't occur anymore.

I also noticed is that each time I make changes to the `destination` vector value in `Properties` I get this error:
```
<09:20:51> [Warning] (Prefab Serialization) - The target for "replace" operation doesn't exist.
    Reason: The operation has halted during a merge operation for '/0' because a field or element has an invalid value.
<09:20:51> [Warning] (Prefab) - Some of the patches were not successfully applied.
```

I also noticed that after saving the level and reloading it the `Properties.destination` was at its `default` value (0, 5, 0)  

I also noticed that this bug occurs with new scripts that have not been loaded before in a level. After the workaround (shown below) the script becomes usable with new entities or new levels, etc. Which possibly means there could be an asset processing issue going on.

### Workaround
I was able to overcome this bug by simply unselecting the lua script `object_relocator.lua`, and selecting again in the Lua Script Component. Then I was able to make changes to  `Properties.destination` and save the level. And when entering into game mode it would work as expected.

**Comment by galibzon** [comment]
This is a lua scripting bug, but seems heavily influenced by the prefab system.

**Comment by carbonated-niko** [comment]
I know this is an older issue - but it appears to still be valid. If I have prefabs that contain a script component, they need to be resaved after a script property is added even if the property has a default value. For some scripts, that can be very tedious as one change can force us to update lots of prefabs.

I would expect that the default value should suffice in situations like this with minimal runtime cost (really just a minor cost on initialization of the component)

**Comment by nick-l-o3de** [comment]
I'm looking into it.  My guess is that default values are not stored in the prefab file (which is by design, to not waste space storing every single property on every single component).  But somehow I suspect on script components that is causing a problem.

**Comment by nick-l-o3de** [comment]
There appears to be actually 2 related problems here.

One, the old system used to write every single value from every single property on every single component to the output file - that means even if values for things that you would never change on a component, for example, rotate 0,0,0 or uniform scale 0, or the thousands of fields on every component, every single one would be in the slice file.

In prefabs, we only write values that differ from "default".  So if all you do is add a component to an entity and leave all its properties as default, it writes nothing but the fact that you added a component.  Which is great for compactness.

Nothing bad so far...

But... it starts to go wrong with dynamic script properties on lua.  The way the system used to work is that since every property is always present, simply creating a "Properties" tag on a lua file would cause the AZ::Entity's ScriptComponent to have a m_properties (C++) object that had a cached value of that property, as a C++ object (so for example, a AZ::Vector3) in a table. 
Whether you changed it from default or not.   When your script then asked for `Properties.blah` it would immediately return the cached c++ object, from the index table function.   The first warning we see warning about 'performance problems' is because those properties are not there anymore.  

So they are issuing a performance warning and falling back to calling the index function on the metatable (which should be the Properties lua table common to all instances of the entity).  This is a performance problem because calling that means for example, actually *interpreting* a string like `AZ::Vector3(0,5,0)` into an actual AZ::Vector3 by constructing it at that moment is obviously a lot more expensive than returning an already-made one.  

However, this is where the second problem comes in, that being, that doesn't work either.  It ultimately returns a `nil`.  So there are two problems here.  

One is that the prefab doesn't contain the default properties - yes, even though technically prefab doesn't write any properties, it should be writing and saving any properties that differ from the default instantation of the type - in other words, if you construct a new, empty EditorScriptComponent, it will only write the fields that differ.  But assigning a script to a EditorScriptComponent does in fact fill in the m_scriptComponent field, which in turn, does fill in the m_properties values with newly constructed defaults.  A script component with properties **does** differ from a default constructed Script Component, so the first place it goes wrong is there...  simply assigning a script that has properties should immediately mark the prefab as dirty, and saving at that point should immediately cause those properties to appear in the prefab file.  They do not.  That's where I'm going to look first.

**Comment by nick-l-o3de** [comment]
Note that while it is tempting to just make the asset load immediately instead of ticks, and that would sort of solve it, tehre is another issue here, and that is that you can modify the Properties table by editing the LUA file itself, in which case, asset will reload but the same problem (data being saved into a prefab that has nothing to do with user modification) becomes an issue.

**Comment by nick-l-o3de** [comment]
Yeah, adding some undo magic in the right place fixes the first problem, and it does so retroactively.  I'd still like it to be possible for it to use the fallback properties table with the perf warning though.

The other issue is, with the fix, if you open a level with bad entities in it (ones missing their property table), its going to auto fix them on load, (creating the property table and marking the entity dirty for save), which will mark the level as modified as soon as you open it.  It sounds good... but..  I'm not sure what this does to situations where the prefab is inside a prefab, it might consider those changes as overrides, which is going to be complex.   I have to find a way to make it ONLY do this when the data isn't already good.

However, I'm not sure there's another fix that can be made, as we also want something to doctor these values when the script is auto reloaded from disk, and has new properties.   I'll do some experiemnting and see if I can get all the edge cases.

**Comment by nick-l-o3de** [comment]
Merged into stabilization - needs merge into dev before close.
